### PR TITLE
Vault

### PR DESCRIPTION
**Pull request by spladug**
This adds support for secrets stored in Hashicorp Vault. What _is_ here should be heavily documented in the diff itself, so I'll defer to that rather than writing more here in the PR body. This is all rather much, so please do tear it apart and let me know if anything is confusing, misleading, dangerous, insecure, etc.

Open questions:

* monitoring of fetcher daemon: The daemon will run as an appcommon::daemon which means we'll get log shipping and QA checks for free. There's not any facility for alerting on sudden failure down the line though. 

Future work:

* [HVAC](https://github.com/ianunruh/hvac) integration. The integration should take the vault_token from this stuff and allow full in-app access to vault's more interesting features (like the transit backend). All with diagnostics instrumentation, of course!
* [Database credentials](https://www.vaultproject.io/docs/secrets/databases/index.html): The database helpers should get integration with the Vault stuff so that apps can automatically fetch rotating DB credentials.

### CODE REVIEW

File: baseplate/secrets/fetcher.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        "expiration": "2017-05-05T09:42.143:35Z",
+        "secrets": {
+            "secret/one": {...},
+            "secret/two": {...},
+            "secret/three": {...}
+        },
+        "vault_token": "9da4241c-3460-11e7-84ac-0e9f9d32522f"
+    }
+
+The vault_token can be used for direct communication with Vault using the
```
Comment (m104) [inline baseplate/secrets/fetcher.py]: Is the vault_token going to be per-instance, app, or role? In other words, can I piggy back off of this token to store cubbyhole data that is accessible across the app?

Comment (spladug) [inline baseplate/secrets/fetcher.py]: Per instance and rotating periodically.

Comment (m104) [inline baseplate/secrets/fetcher.py]: great, thanks!

File: baseplate/secrets/store.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+    """
+
+    def __init__(self, path):
+        self._path = path
+        self._expiration = datetime.datetime.min
+        self._secrets = None
+        self._vault_token = None
+
+    def _load_if_expired(self):
```
Comment (spladug) [inline baseplate/secrets/store.py]: Should we force a reload if the mtime is higher than when we last read the file even if not expired yet? That would mean extra `stat` calls, but that should be in-cache most of the time. The upside would be that we are sure we get the latest copy if someone restarts the fetcher (e.g. for new configs). Alternatively, we punt and say "if you really need new data, send an app restart".

Comment (ckwang8128) [inline baseplate/secrets/store.py]: I think it makes sense to force a reload if we notice "newness"; a big part of this is abstracting the process of secret discovery from the app, needing to force a restart for cases like new configs feels antithetical to that goal. The only case where I'd be against is, like you pointed out, if there's major performance hits from the extra `stat`s.

**Comment by ckwang8128** [comment]
@:fish: from a code and structure perspective. Non-blocking question: How does secret revocation look like in this pipeline, specifically with respect to the VersionedSecret implementation here (this is from the perspective of someone who doesn't know a ton about Vault details)? Will the `previous` attribute on the secret just be expected to change/be removed if a secret is explicitly removed and replaced?

**Comment by spladug** [comment]
> How does secret revocation look like in this pipeline, specifically with respect to the VersionedSecret implementation here ... Will the previous attribute on the secret just be expected to change/be removed if a secret is explicitly removed and replaced?

Regular rotation would be like 

* read secret
* `new_value["previous"] = old_value["current"]`
* `new_value["current"] = old_value["next"]`
* `new_value["next"] = generate_new_secret()`
* write secret.

Revocation would depend on how seriously we need to revoke: in a hard cutover situation, it'd just be replacing the dictionary with `current = generate_new_secret()` and let the other two be null.

**Comment by ckwang8128** [comment]
SGTM, just making sure I understand the process!

File: baseplate/crypto.py [hunk 3]
```diff
@@ -92,89 +122,97 @@ class SignatureInfo(_SignatureInfo):
...
-            message, max_age=datetime.timedelta(days=1))
+    """
+    version = 1
+    expiration = int(time.time() + max_age.total_seconds())
+    header = _HEADER_FORMAT.pack(version, expiration)
+    digest = _compute_digest(secret.current, header, message)
+    return base64.urlsafe_b64encode(header + digest)
+
+
+def validate_signature(secret, message, signature):
```
Comment (JordanMilne) [inline baseplate/crypto.py]: Probably best saved for another PR, but it'd be nice to have a wrapper like this around cryptography.io's [`MultiFernet`](https://cryptography.io/en/latest/fernet/#cryptography.fernet.MultiFernet) that can accept a `VersionedSecret`

File: baseplate/crypto.py [hunk 4]
```diff
@@ -92,89 +122,97 @@ class SignatureInfo(_SignatureInfo):
...
+    :param datetime.timedelta max_age: The amount of time in the future
+        the signature will be valid for.
+    :return: An encoded signature.
 
-        signer = MessageSigner(b"supersecret")
-        signature = signer.make_signature(
-            message, max_age=datetime.timedelta(days=1))
+    """
+    version = 1
+    expiration = int(time.time() + max_age.total_seconds())
```
Comment (spladug) [inline baseplate/crypto.py]: Mayyyybe, but if so I'd want you to have to opt into it really noisily. Trying to make it hard to do the wrong thing in the helpers that go into this lib. Do you have any use cases in mind? (And perhaps let's keep that for another PR since this is just reworking the helpers to use versioned secrets?)

File: baseplate/crypto.py (hunk 3 above)
Comment (spladug) [inline baseplate/crypto.py]: Yeah! Would love to get more helpers in that take advantage of the versioned things.

File: baseplate/secrets/store.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+    def __str__(self):  # pragma: nocover
+        return "{}: {}".format(self.path, self.message)
+
+
+_VersionedSecret = collections.namedtuple(
+    "VersionedSecret", "previous current next")
+
+
+class VersionedSecret(_VersionedSecret):
```
Comment (spladug) [inline baseplate/secrets/store.py]: Negative. This is purely our convention. I'd like to get a cron job that rotates keys going but that'd be vault-side so apps don't even need to know it exists as long as they can handle versioned secrets OK.

File: baseplate/secrets/fetcher.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        response.raise_for_status()
+        payload = response.json()
+        return payload["data"], ttl_to_time(payload["lease_duration"])
+
+
+def octal_integer(text):
+    return int(text, base=8)
+
+
+def user_name(text):
```
Comment (spladug) [inline baseplate/secrets/fetcher.py]: Hm, yeah, it looks a bit weird when doing it like this but I'm going for the declarative style of the rest of the config helpers. Maybe I should capitalize like they do? or just ditch it?

File: baseplate/crypto.py (hunk 3 above)
Comment (spladug) [inline baseplate/crypto.py]: Also thanks for linking that. It makes me feel better to see their implementation of multi-version on the read path is the same. I was a bit worried there'd be some timing attack concern I wasn't thinking about.

File: baseplate/secrets/store.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        :rtype: :py:class:`VersionedSecret`
+
+        """
+        secret_attributes = self.get_raw(path)
+
+        if secret_attributes.get("type") != "versioned":
+            raise CorruptSecretError(path, "secret does not have type=versioned")
+
+        previous = secret_attributes.get("previous")
+        next = secret_attributes.get("next")
```
Comment (spladug) [inline baseplate/secrets/store.py]: Yes, yucky, thanks. Fixed.

File: baseplate/secrets/fetcher.py (hunk 6 above)
Comment (spladug) [inline baseplate/secrets/fetcher.py]: Renamed them to be clearer.

**Comment by spladug** [comment]
(rebased to deal with merge conflicts. changes since first review pass are still in the lower-case fixup commits)

File: baseplate/secrets/store.py (hunk 2 above)
Comment (spladug) [inline baseplate/secrets/store.py]: OK, done!

**Comment by spladug** [comment]
:haircut_man:

File: baseplate/secrets/store.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    cache or store the values returned by this class's methods but rather get
+    them from this class each time you need them. The secrets are served from
+    memory so there's little performance impact to doing so and you will be
+    sure to always have the current version in the face of key rotation etc.
+
+    """
+
+    def __init__(self, path):
+        self._path = path
+        self._expiration = datetime.datetime.min
```
Comment (dellis23) [inline baseplate/secrets/store.py]: Wow, TIL about `datetime.datetime.min`

File: baseplate/secrets/store.py (hunk 2 above)
Comment (dellis23) [inline baseplate/secrets/store.py]: Obviously late to this, but yeah, I like the idea of checking `mtime` but I wonder if this would get funky with a service that starts to need to check secrets many thousands of times per second.  We could do a check at most once per second or something like that.  I wonder if that's unrealistic though.  Would secrets mostly be grabbed for external calls and it be unlikely that we'd be grabbing them that frequently?  I'd bet it'd be cached in the kernel and would be a cheap call though.  Just thinking out loud.  Probably all pre-mature optimization.

File: baseplate/secrets/store.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        self._secrets = None
+        self._vault_token = None
+
+    def _load_if_expired(self):
+        if self._expiration < datetime.datetime.utcnow():
```
Comment (dellis23) [inline baseplate/secrets/store.py]: I haven't seen the later commit yet, but I wonder if, since we're presumably doing the mtime method, we even need to also do it this way!  Would make the code cleaner at least.

File: baseplate/secrets/store.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+            with open(self._path) as f:
+                raw_data = json.load(f)
+
+            self._expiration = datetime.datetime.strptime(
+                raw_data["expiration"], ISO_FORMAT)
+            self._vault_token = raw_data["vault_token"]
+            self._secrets = raw_data["secrets"]
+
+    def make_object_for_context(self, name, server_span):  # pragma: nocover
```
Comment (dellis23) [inline baseplate/secrets/store.py]: What's up with this?

File: baseplate/secrets/store.py (hunk 9 above)
Comment (spladug) [inline baseplate/secrets/store.py]: Hmm, that's a good point. There's no real need to keep doing the expiration check if the mtime is being checked: even if the secrets _are_ expired, if the mtime hasn't changed we're not gonna magically get new secrets. I'll tear the expiration stuff out. Thanks!

File: baseplate/secrets/store.py (hunk 10 above)
Comment (spladug) [inline baseplate/secrets/store.py]: It lets you do

```
baseplate = Baseplate()
secrets = SecretsStore(...)
baseplate.add_to_context("secrets", secrets)
```

and then you get the secrets store on your context object. Similar to how the event queues work.

File: baseplate/secrets/store.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        Versioned secrets are a convention of key/value pairs in the raw secret
+        payload. The following keys are significant:
+
+        ``type``
+            This must always be ``versioned`` for this method.
+        ``current``, ``next``, and ``previous``
+            The raw secret value's versions. ``current`` is the "active"
+            version, which is used for new creation/signing operations.
+            ``previous`` and ``next`` are only used for validation (e.g.
+            checking signatures) to ensure continuity when keys rotate. Both
```
Comment (dellis23) [inline baseplate/secrets/store.py]: Hmm, maybe you can explain this to me in person, but I don't understand what the usefulness of keeping the previous version is if it's not actually valid for doing the operation (i.e. I'm assuming you mean it instantly expires when rotation happens.  If the client for whatever reason hasn't gotten that update yet, does the previous token then expire as well?

File: baseplate/secrets/store.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+            ``previous`` and ``next`` are optional.
+        ``encoding``
+            (Optional) If present, how to decode the values from how they are
+            encoded at rest (Only ``base64`` currently supported).
```
Comment (dellis23) [inline baseplate/secrets/store.py]: super super nit: that capital `Only`.

File: baseplate/secrets/store.py (hunk 11 above)
Comment (spladug) [inline baseplate/secrets/store.py]: Check out how it's used in the message signer for when the previous/next values come into play.

File: baseplate/secrets/store.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        if secret_attributes.get("type") != "versioned":
+            raise CorruptSecretError(path, "secret does not have type=versioned")
+
+        previous = secret_attributes.get("previous")
+        next = secret_attributes.get("next")
+
+        try:
+            current = secret_attributes["current"]
+        except KeyError as exc:
+            raise CorruptSecretError(path, "secret does not have %s" % exc)
```
Comment (dellis23) [inline baseplate/secrets/store.py]: I could see this error being confusing `secret does not have current`.  Maybe explicitly specifying that it's the key missing there that caused the error to be raised?

File: baseplate/secrets/store.py (hunk 9 above)
Comment (spladug) [inline baseplate/secrets/store.py]: Done.

File: baseplate/secrets/store.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+        """
+        return cls(
+            previous=None,
+            current=value,
+            next=None,
+        )
+
+
+def _decode_secret(path, encoding, value):
+    if encoding == "identity":
```
Comment (dellis23) [inline baseplate/secrets/store.py]: Why is this called `identity`?  Is it used for something special that makes it tend to not be random bytes that need to be encoded?  Also, I've noticed it is the default if no encoding is specified, so it almost feels like just a default non-encoding?  I guess I'm just not sure what it's for.

File: baseplate/secrets/store.py (hunk 10 above)
Comment (spladug) [inline baseplate/secrets/store.py]: Added a docstring.

File: baseplate/secrets/store.py (hunk 12 above)
Comment (spladug) [inline baseplate/secrets/store.py]: it's called big-O notation for a reason! 

(fixed)

File: baseplate/secrets/store.py (hunk 14 above)
Comment (spladug) [inline baseplate/secrets/store.py]: Yup, the identity encoding. Like multiplying by 1.

> (algebra, computing) Any function which maps all elements of its domain to themselves.

File: docs/baseplate/secrets.rst [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+--------------
+
+The secret fetcher is a sidecar that is run as a single daemon on each server.
+It authenticates to Vault as the server itself and gets appropriate policies
+for access to secrets accordingly. Once authenticated, it fetches a given list
+of secrets from Vault and stores all of the data in a local file. It will
+automatically re-fetch secrets as their leases expire, ensuring that key
+rotation happens on schedule.
+
+Because this is a sidecar, individual application processes don't need to talk
```
Comment (dellis23) [inline docs/baseplate/secrets.rst]: Cool design btw.

File: baseplate/secrets/store.py (hunk 13 above)
Comment (spladug) [inline baseplate/secrets/store.py]: GP on the exception message being unhelpful. And whoops, that weird parameterization was from an earlier version. Cleaned up.

File: baseplate/secrets/store.py (hunk 11 above)
Comment (dellis23) [inline baseplate/secrets/store.py]: Ahh yes, thanks.  Makes sense.

File: baseplate/secrets/store.py [hunk 16]
```diff
@@ -244,6 +241,18 @@ def get_versioned(self, path):
             next=next_value and _decode_secret(path, encoding, next_value),
         )
 
+    def make_object_for_context(self, name, server_span):  # pragma: nocover
+        """Return an object that can be added to the context object.
```
Comment (dellis23) [inline baseplate/secrets/store.py]: Thanks!

**Comment by dellis23** [comment]
🐟 Thanks for the overview

**Comment by ckwang8128** [comment]
:fish:
